In [1]:
import pandas as pd
import numpy as np

In [2]:
test_df = pd.read_csv("https://raw.githubusercontent.com/WHPAN0108/BHT-DataScience-S23/main/classification/data/Assigment/aug_test.csv")
test_df.head()

city_development_index  gender      relevent_experience  \
0                   0.624     NaN  Has relevent experience   
1                   0.920  Female   No relevent experience   
2                   0.767     NaN  Has relevent experience   
3                   0.910    Male   No relevent experience   
4                   0.624    Male  Has relevent experience   

  enrolled_university education_level major_discipline experience  \
0    Full time course        Graduate            Other          3   
1       no_enrollment        Graduate             STEM          5   
2    Full time course        Graduate             STEM         10   
3                 NaN     High School              NaN         10   
4    Part time course        Graduate             STEM          3   

          company_type last_new_job  training_hours  target  
0              Pvt Ltd            1             134       0  
1  Early Stage Startup            1              34       1  
2              Pvt Ltd            2              90       0  
3                  NaN        never              42       0  
4              Pvt Ltd            1             198       0

In [3]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   city_development_index  100 non-null    float64
 1   gender                  72 non-null     object 
 2   relevent_experience     100 non-null    object 
 3   enrolled_university     96 non-null     object 
 4   education_level         98 non-null     object 
 5   major_discipline        88 non-null     object 
 6   experience              100 non-null    object 
 7   company_type            64 non-null     object 
 8   last_new_job            100 non-null    object 
 9   training_hours          100 non-null    int64  
 10  target                  100 non-null    int64  
dtypes: float64(1), int64(2), object(8)
memory usage: 8.7+ KB


In [4]:
#entfernen der fehlenden Werte aus den testdaten
test_df = test_df.dropna()

In [5]:
test_df.describe()

city_development_index  training_hours     target
count               43.000000       43.000000  43.000000
mean                 0.858093       69.930233   0.139535
std                  0.105406       62.299735   0.350605
min                  0.558000        4.000000   0.000000
25%                  0.829500       22.500000   0.000000
50%                  0.910000       50.000000   0.000000
75%                  0.920000       98.500000   0.000000
max                  0.939000      226.000000   1.000000

In [6]:
test_df.shape

(43, 11)

In [7]:
train_df = pd.read_csv("https://raw.githubusercontent.com/WHPAN0108/BHT-DataScience-S23/main/classification/data/Assigment/aug_train.csv")
train_df.head()

city_development_index  gender      relevent_experience  \
0                   0.624    Male   No relevent experience   
1                   0.926    Male  Has relevent experience   
2                   0.920    Male  Has relevent experience   
3                   0.624    Male   No relevent experience   
4                   0.920  Female  Has relevent experience   

  enrolled_university education_level major_discipline experience  \
0       no_enrollment     High School              NaN          5   
1       no_enrollment        Graduate             STEM        >20   
2       no_enrollment        Graduate             STEM        >20   
3    Full time course     High School              NaN          1   
4       no_enrollment         Masters             STEM        >20   

    company_type last_new_job  training_hours  target  
0            NaN        never              21       0  
1            NaN           >4              12       0  
2  Public Sector           >4              26       0  
3            NaN        never              30       1  
4            NaN           >4              46       0

In [8]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2100 entries, 0 to 2099
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   city_development_index  2100 non-null   float64
 1   gender                  1585 non-null   object 
 2   relevent_experience     2100 non-null   object 
 3   enrolled_university     2051 non-null   object 
 4   education_level         2049 non-null   object 
 5   major_discipline        1768 non-null   object 
 6   experience              2090 non-null   object 
 7   company_type            1415 non-null   object 
 8   last_new_job            2048 non-null   object 
 9   training_hours          2100 non-null   int64  
 10  target                  2100 non-null   int64  
dtypes: float64(1), int64(2), object(8)
memory usage: 180.6+ KB


In [9]:
##entfernen der fehlenden Werte aus den trainigsdaten
train_df = train_df.dropna()

In [10]:
train_df.describe()

city_development_index  training_hours      target
count              981.000000      981.000000  981.000000
mean                 0.847670       66.977574    0.172273
std                  0.114722       58.955825    0.377810
min                  0.493000        2.000000    0.000000
25%                  0.802000       24.000000    0.000000
50%                  0.910000       50.000000    0.000000
75%                  0.920000       90.000000    0.000000
max                  0.949000      336.000000    1.000000

In [11]:
train_df.shape

(981, 11)

## Task1 Data clean, imputation

1.in experience, replace >20 to 21; <1 to 1, and convert this as a numerical column

In [12]:
## Ersetzen ">20" durch "21" im trainingsdatensatz
train_df["experience"]= train_df["experience"].replace(">20", "21")

## Ersetzen "<1" durch "1" im trainingsdatensatz
train_df["experience"]= train_df["experience"].replace("<1", "1")

## Konvertieren der Spalte in numerischen Datentyp
train_df["experience"]= pd.to_numeric(train_df["experience"])

In [13]:
print(train_df['experience'].dtype)

int64


In [14]:
## Ersetzen ">20" durch "21" im testdatensatz
test_df["experience"]= test_df["experience"].replace(">20", "21")

## Ersetzen "<1" durch "1" im testdatensatz
test_df["experience"]= test_df["experience"].replace("<1", "1")

## Konvertieren der Spalte in numerischen Datentyp
test_df["experience"]= pd.to_numeric(test_df["experience"])

In [15]:
print(test_df['experience'].dtype)

int64


2. in last_new_job, replace >4 to 5; never to 0, and convert this as a numerical column

In [16]:
## Ersetzen ">4" durch "5" im trainingsdatensatz
train_df["last_new_job"]= train_df["last_new_job"].replace(">4", "5")
#### Ersetzen ">never" durch "0" im trainingsdatensatz
train_df["last_new_job"]= train_df["last_new_job"].replace("never", "0")
#### Konvertieren der Spalte in numerischen Datentyp
train_df["last_new_job"]= pd.to_numeric(train_df["last_new_job"])

In [17]:
## Ersetzen ">4" durch "5" im testdatensatz
test_df["last_new_job"]= test_df["last_new_job"].replace(">4", "5")
#### Ersetzen ">never" durch "0" im testdatensatz
test_df["last_new_job"]= test_df["last_new_job"].replace("never", "0")
## Konvertieren der Spalte in numerischen Datentyp
test_df["last_new_job"]= pd.to_numeric(test_df["last_new_job"])

3. If the column is categorical, impute the missing value as its mode. If the column is numerical, impute the missing value as its median

In [18]:
#Aufbereitung für trainingdatensatz

# Finde Spalten mit fehlenden Werten und deren Datentypen
missing_columns = train_df.columns[train_df.isnull().any()]
column_types = train_df[missing_columns].dtypes

## Imputiere fehlende Werte in kategorischen Spalten mit Modus
for column in missing_columns[column_types == "object"]:
    train_df[column].fillna(train_df[column].mode()[0], inplace=True)

    # Impute missing values in numerical columns with median
for column in missing_columns[column_types != "object"]:
    train_df[column].fillna(train_df[column].median(), inplace=True)


In [19]:
#Aufbereitung für testdatensatz

# Finde Spalten mit fehlenden Werten und deren Datentypen
missing_columns = test_df.columns[test_df.isnull().any()]
column_types = test_df[missing_columns].dtypes

## Imputiere fehlende Werte in kategorischen Spalten mit Modus
for column in missing_columns[column_types == "object"]:
    test_df[column].fillna(train_df[column].mode()[0], inplace=True)

    # Impute missing values in numerical columns with median
for column in missing_columns[column_types != "object"]:
    test_df[column].fillna(test_df[column].median(), inplace=True)

## Task2 Classification

In [20]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.compose import ColumnTransformer
from sklearn.metrics import mean_squared_error
from sklearn import preprocessing


In [21]:
from sklearn.preprocessing import LabelEncoder


Im folgenden wird  Imputation fehlender Werte in der Spalte  des DataFrames train_df und test_df durchgeführt und anschließend eine Label-Encoding-Transformation auf die imputierten Werte angewendet. Dieser Code ermöglicht es, fehlende Werte in den Spalten zu behandeln und sie in numerische Werte umzuwandeln, um sie für weitere Analysen oder das Training von Machine-Learning-Modellen zu verwenden.

In [23]:
train_df['gender'].fillna(method='ffill',inplace=True)
genderEncoder = LabelEncoder().fit(train_df['gender'])
gender_List = genderEncoder.classes_
train_df['gender'] = genderEncoder.transform(train_df['gender'])

test_df['gender'].fillna(method='ffill',inplace=True)
genderEncoder = LabelEncoder().fit(test_df['gender'])
gender_List = genderEncoder.classes_
test_df['gender'] = genderEncoder.transform(test_df['gender'])

In [24]:
train_df['enrolled_university'] = train_df['enrolled_university'].fillna(train_df['enrolled_university'].value_counts().index[0])
EnrolledUniversityEncoder = LabelEncoder().fit(train_df['enrolled_university'])
EnrolledUniversityEncoder_List = EnrolledUniversityEncoder.classes_
train_df['enrolled_university'] = EnrolledUniversityEncoder.transform(train_df['enrolled_university'])

test_df['enrolled_university'] = test_df['enrolled_university'].fillna(test_df['enrolled_university'].value_counts().index[0])
EnrolledUniversityEncoder = LabelEncoder().fit(test_df['enrolled_university'])
EnrolledUniversityEncoder_List = EnrolledUniversityEncoder.classes_
test_df['enrolled_university'] = EnrolledUniversityEncoder.transform(test_df['enrolled_university'])

In [25]:
train_df['education_level'].fillna(method='ffill',inplace=True)
EducationLevelEncoder = LabelEncoder().fit(train_df['education_level'])
EducationLevelEncoder_List = EducationLevelEncoder.classes_
train_df['education_level'] = EducationLevelEncoder.transform(train_df['education_level'])

test_df['education_level'].fillna(method='ffill',inplace=True)
EducationLevelEncoder = LabelEncoder().fit(test_df['education_level'])
EducationLevelEncoder_List = EducationLevelEncoder.classes_
test_df['education_level'] = EducationLevelEncoder.transform(test_df['education_level'])


In [26]:
train_df['major_discipline'].fillna(method='ffill',inplace=True)
major_disciplineEncoder = LabelEncoder().fit(train_df['major_discipline'])
major_disciplineEncoder_List = major_disciplineEncoder.classes_
train_df['major_discipline'] = major_disciplineEncoder.transform(train_df['major_discipline'])

test_df['major_discipline'].fillna(method='ffill',inplace=True)
major_disciplineEncoder = LabelEncoder().fit(test_df['major_discipline'])
major_disciplineEncoder_List = major_disciplineEncoder.classes_
test_df['major_discipline'] = major_disciplineEncoder.transform(test_df['major_discipline'])

In [27]:
train_df['experience'].fillna(method='ffill',inplace=True)
experienceEncoder = LabelEncoder().fit(train_df['experience'])
experienceEncoder_List = experienceEncoder.classes_
train_df['experience'] = experienceEncoder.transform(train_df['experience'])

test_df['experience'].fillna(method='ffill',inplace=True)
experienceEncoder = LabelEncoder().fit(test_df['experience'])
experienceEncoder_List = experienceEncoder.classes_
test_df['experience'] = experienceEncoder.transform(test_df['experience'])

In [28]:
train_df['company_type'].fillna(method='ffill',inplace=True)
company_typeEncoder = LabelEncoder().fit(train_df['company_type'])
company_typeEncoder_List = company_typeEncoder.classes_
train_df['company_type'] = company_typeEncoder.transform(train_df['company_type'])

test_df['company_type'].fillna(method='ffill',inplace=True)
company_typeEncoder = LabelEncoder().fit(test_df['company_type'])
company_typeEncoder_List = company_typeEncoder.classes_
test_df['company_type'] = company_typeEncoder.transform(test_df['company_type'])

In [29]:
train_df['last_new_job'].fillna(method='ffill',inplace=True)
last_new_jobEncoder = LabelEncoder().fit(train_df['last_new_job'])
last_new_jobEncoder_List = last_new_jobEncoder.classes_
train_df['last_new_job'] = last_new_jobEncoder.transform(train_df['last_new_job'])

test_df['last_new_job'].fillna(method='ffill',inplace=True)
last_new_jobEncoder = LabelEncoder().fit(test_df['last_new_job'])
last_new_jobEncoder_List = last_new_jobEncoder.classes_
test_df['last_new_job'] = last_new_jobEncoder.transform(test_df['last_new_job'])

In [30]:
train_df['relevent_experience'].fillna(method='ffill',inplace=True)
relevent_experienceEncoder = LabelEncoder().fit(train_df['relevent_experience'])
relevent_experience_List = relevent_experienceEncoder.classes_
train_df['relevent_experience'] = relevent_experienceEncoder.transform(train_df['relevent_experience'])

test_df['relevent_experience'].fillna(method='ffill',inplace=True)
relevent_experienceEncoder = LabelEncoder().fit(test_df['relevent_experience'])
relevent_experience_List = relevent_experienceEncoder.classes_
test_df['relevent_experience'] = relevent_experienceEncoder.transform(test_df['relevent_experience'])

Überprüfung :

In [31]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 981 entries, 2 to 2099
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   city_development_index  981 non-null    float64
 1   gender                  981 non-null    int64  
 2   relevent_experience     981 non-null    int32  
 3   enrolled_university     981 non-null    int32  
 4   education_level         981 non-null    int32  
 5   major_discipline        981 non-null    int32  
 6   experience              981 non-null    int64  
 7   company_type            981 non-null    int32  
 8   last_new_job            981 non-null    int64  
 9   training_hours          981 non-null    int64  
 10  target                  981 non-null    int64  
dtypes: float64(1), int32(5), int64(5)
memory usage: 72.8 KB


In [32]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43 entries, 1 to 99
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   city_development_index  43 non-null     float64
 1   gender                  43 non-null     int64  
 2   relevent_experience     43 non-null     int32  
 3   enrolled_university     43 non-null     int32  
 4   education_level         43 non-null     int32  
 5   major_discipline        43 non-null     int32  
 6   experience              43 non-null     int64  
 7   company_type            43 non-null     int32  
 8   last_new_job            43 non-null     int64  
 9   training_hours          43 non-null     int64  
 10  target                  43 non-null     int64  
dtypes: float64(1), int32(5), int64(5)
memory usage: 3.2 KB


1. Build a classification model from the training set ( you can use any algorithm)

In [33]:
# Aufteilen der Features (X) und der Zielvariable (y)
X_train = train_df.drop("target", axis=1) # Features
y_train= train_df["target"]  # Zielvariable

In [34]:
# Separate Aufteilen der Features (X) und der Zielvariable (y) für test
X_test = test_df.drop("target", axis=1)  # Features für test
y_test = test_df["target"]  # zielvariable für Test


In [35]:
# Initialisieren des Random Forest Klassifikators
clf = RandomForestClassifier()

In [36]:
clf.fit(X_train, y_train)

RandomForestClassifier()

2.generate the confusion matrix for the training set and calculate the accuracy, precision, recall, and F1-score 

In [37]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

In [38]:
# Vorhersagen für die Trainingsmenge treffen
y_pred_train = clf.predict(X_train)

# Erzeugen Sie die Konfusionsmatrix für die Trainingsmenge
cm_train = confusion_matrix(y_train, y_pred_train)
print("Confusion Matrix (Trainingmenge):")
print(cm_train)

# Berechnen Sie die Genauigkeit für die Trainingsmenge
accuracy_train = accuracy_score(y_train, y_pred_train)
print("Accuracy (Trainingmenge):", accuracy_train)

# Berechnen Sie die Präzision für den Trainingssatz
precision_train = precision_score(y_train, y_pred_train)
print("Precision (Trainingmenge):", precision_train)

#Berechnung der Recall für die Trainingsmenge
recall_train = recall_score(y_train, y_pred_train)
print("Recall (Trainingmenge):", recall_train)

# Berechnen des F1-Scores für den Trainingssatz
f1_train = f1_score(y_train, y_pred_train)
print("F1-score (Trainingmenge):", f1_train)

Confusion Matrix (Trainingmenge):
[[812   0]
 [  0 169]]
Accuracy (Trainingmenge): 1.0
Precision (Trainingmenge): 1.0
Recall (Trainingmenge): 1.0
F1-score (Trainingmenge): 1.0


3. Applying the model in the test set and generating the prediction

In [39]:
# Vorhersagen für die Testmenge treffen
y_pred_test = clf.predict(X_test)

print("Predictions:", y_pred_test)

Predictions: [0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0]


4. generate the confusion matrix from the test set and calculate the accuracy, precision, recall, and F1-score

In [40]:
#Erzeugen Sie die Konfusionsmatrix für die Testmenge
cm_test = confusion_matrix(y_test, y_pred_test)
print("Confusion Matrix (Testmenge):")
print(cm_test)

# Berechnen Sie die Genauigkeit für die Testsmenge
accuracy_test = accuracy_score(y_test, y_pred_test)
print("Accuracy (Testmenge):", accuracy_test)

# Berechnen Sie die Präzision für den Testmenge
precision_test = precision_score(y_test, y_pred_test)
print("Precision (Testmenge):", precision_test)

# Berechnung von Recall für die Testsmenge
recall_test = recall_score(y_test, y_pred_test)
print("Recall (Testmenge):", recall_test)

# Berechnen des F1-Scores für den Trainingssatz
f1_test = f1_score(y_test, y_pred_test)
print("F1-score (Testmenge):", f1_test)


Confusion Matrix (Testmenge):
[[35  2]
 [ 4  2]]
Accuracy (Testmenge): 0.8604651162790697
Precision (Testmenge): 0.5
Recall (Testmenge): 0.3333333333333333
F1-score (Testmenge): 0.4


5. compare the results between the training and test set

In [41]:
print("Evaluation der Trainingsmenge:")
print("Accuracy:", accuracy_train)
print("Precision:", precision_train)
print("Recall:", recall_train)
print("F1-score:", f1_train)
print()

print("Evaluation der Testmenge:")
print("Accuracy:", accuracy_test)
print("Precision:", precision_test)
print("Recall:", recall_test)
print("F1-score:", f1_test)

Evaluation der Trainingsmenge:
Accuracy: 1.0
Precision: 1.0
Recall: 1.0
F1-score: 1.0

Evaluation der Testmenge:
Accuracy: 0.8604651162790697
Precision: 0.5
Recall: 0.3333333333333333
F1-score: 0.4


Die gegebenen Ergebnisse zeigen eine deutliche Diskrepanz zwischen der Leistung des Modells auf der Trainingsmenge und der Testmenge.
Die Ergebnisse der Trainigsmenge zeigen, dass das Model perfekte Leistung erzielt. Das zeigt der Wert 1.0 bei Accuracy, Precision, Recall und F1-Score, und bedeutet , dass das Modell alle Beispiele richtig klassifiziert hat.
Im Gegensatz dazu zeigen die Ergebnisse der Testmenge, dass das Modell etwa 86% der Beispiele korrekt klassifiziert hat. Die Precision liegt bei 0.5, was darauf hindeutet, dass von den als positiv vorhergesagten Beispielen etwa 50% tatsächlich positiv waren. Der Recall beträgt 0.333, was bedeutet, dass das Modell nur etwa 33,3% der tatsächlich positiven Beispiele erfasst hat. Der F1-Score, der das Gleichgewicht zwischen Precision und Recall darstellt, beträgt 0.444.
Der deutliche Unterschied zwieschen Leistungen der Trainings und Testmenge deutet darauf hin, dass das Modell Schwierigkeiten hat, die in den Trainingsdaten gelernten Muster auf neue Daten anzuwenden. Dies kann ein hinweis auf Overfitting (Überangepastheit) des Models hindeuten.
Das Anpassen von Hyperparametern, die Verwendung von mehr Daten oder die Anwendung von Regularisierungstechniken könnte dabei helfen bessere Generalisierung zu erreichen